### DISCUSSION TITLE
Cloudflare d1

### DISCUSSION THREAD

**Discussion by thunderbug1** [alfateam/rdb#114] [Q&A] [answered]
Hello,

The project is compatible with Sqlite, is it also compatible with Cloudflare d1 which is based on Sqlite?

**Comment by lroal** [comment]
In theory, it should work as long as you run cloudflare with [nodejs compatability flag ](https://developers.cloudflare.com/workers/configuration/compatibility-flags/#nodejs-compatibility-flag)(because it needs nodejs domain)
But I need to create an adapter for it first.

**Comment by thunderbug1** [comment]
thank you very much for your fast answer and the project in general.

Are you planning on building an adapter for d1? I see quite a long list of adapters in the roadmap drafts.

**Reply by lroal** [reply]
Yes, because it seems to be pretty easy. I can prioritize d1 before the others.
The only drawback is lack of unit testing - because it is not possible to run it locally (in a sane way). But I can live with that.

**Reply by thunderbug1** [reply]
oh that's cool thanks.

It actually is possible to run it locally. [Minflare](https://miniflare.dev/) basically just wraps a local sqlite file and acts like d1.
I initialize the local database with ` npx wrangler d1 execute db-name-here --local --file=./schema.sql` where schema.sql creates all the tables.

After this I have a local folder like:
![image](https://github.com/user-attachments/assets/5fb4e52d-89c9-46ee-b3d8-9959bf9b2ee2)

With this database I can interact via the d1 binding but also via any sqlite tools since it is just a sqliteDB file.

I'm not sure if this can be used for unit testing here.

**Reply by lroal** [reply]
I have published a beta version that should be working.
```bash
npm install orange-orm@4.5.0-beta.0
```
I have only tested with miniflare - not the real thing.
Can you verify that it works ?
I guess one needs to bundle it to run on d1.
Remarks that transactions are not supported as they are not supported on d1.
Usage (unbundled):
```js
import map from './map';

export default {
  async fetch(request, env) {
    // If you did not use `DB` as your binding name, change it here
    const db = map.d1(env.DB);
    const orders = await db.order.getAll();
    //....
   //....
  },
};
```

**Reply by lroal** [reply]
Remember to set the compatability flags in toml file to make sure domain works.
```toml
compatibility_flags = [ "nodejs_compat" ]
```

**Reply by lroal** [reply]
Perhaps you need this too, not sure
```toml
compatibility_date = "2024-09-23"
```

**Reply by lroal** [reply]
I have tested it with d1 and it throws error. I need to investigate this.
 ```error
Cannot create pool for database outside node
 ```

**Reply by thunderbug1** [reply]
ok thank you for the updates.

**Reply by lroal** [reply]
FYI:
To make it compatible with Cloudflare D1, I need to refactor some internal components and remove the deprecated [Domain ](https://nodejs.org/api/domain.html) module in Node.js. Currently, I use it internally as an "async context" or "thread singleton" to avoid passing the connection around. Once this change is implemented, it will pave the way for compatibility with other runtimes and databases like Deno and pgLite. This is a necessary internal update and will not affect the public API.

While I can’t provide an exact delivery date, my goal is to complete it by January 2025.

**Reply by thunderbug1** [reply]
ok, thank you.
I ended up using this now: https://www.kysely.dev/

**Comment by lroal** [comment, marked as answer]
Cloudflare is now supported from v.4.5.0
<sub>📄 wrangler.toml</sub>  
```toml
name = "d1-tutorial"
main = "src/index.ts"
compatibility_date = "2025-02-04"

# Bind a D1 database. D1 is Cloudflare’s native serverless SQL database.
# Docs: https://developers.cloudflare.com/workers/wrangler/configuration/#d1-databases
[[d1_databases]]
binding = "DB"
database_name = "<your-name-for-the-database>"
database_id = "<your-guid-for-the-database>"
```

<sub>📄 src/index.ts</sub>  
```javascript
import map from './map';

export interface Env {
  // Must match the binding name in wrangler.toml  
  DB: D1Database;
}

export default {
  async fetch(request, env): Promise<Response> {
    const db = map.d1(env.DB);
    const customers = await db.customer.getAll();
    return Response.json(customers);
  },
} satisfies ExportedHandler<Env>;
```
